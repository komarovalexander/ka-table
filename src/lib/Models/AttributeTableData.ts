import { HTMLAttributes } from 'react';

import { ChildProps, DispatchFunc } from '../types';

export class AttributeTableData {
  public baseFunc!: any;
  public childElementAttributes!: HTMLAttributes<HTMLElement>;
  public childProps!: ChildProps;
  public dispatch!: DispatchFunc;
}
