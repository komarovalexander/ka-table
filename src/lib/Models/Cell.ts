/**
 * Describes the position of cell in table
 */
export class Cell {
  /*
  * Specifies the field of data
  */
  public columnKey!: string;

  /*
  * Specifies the value of row's key where cell is placed
  */
  public rowKey!: any;
}
