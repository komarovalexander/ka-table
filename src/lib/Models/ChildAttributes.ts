
import { ITableAllProps } from '../';
import { ICellContentProps } from '../Components/CellContent/CellContent';
import { ChildAttributesItem } from '../types';

export class ChildAttributes {
  public cell?: ChildAttributesItem<ICellContentProps>;
  public table?: ChildAttributesItem<ITableAllProps>;
}
