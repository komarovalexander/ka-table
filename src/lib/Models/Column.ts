import { DataType, SortDirection } from '../enums';
import {
  CellFunc, EditorFunc, FilterRowFunc, FormatFunc, HeaderCellFunc, SearchFunc, ValidationFunc,
} from '../types';

/**
 * Describes column of table its look and behaviour
 */
export class Column {
  public cell?: CellFunc;
  public dataType?: DataType;
  public editor?: EditorFunc;
  public filterRowCell?: FilterRowFunc;
  public filterRowValue?: any;
  public filterRowOperator?: any;
  public field?: string;
  public format?: FormatFunc;
  public headCell?: HeaderCellFunc;
  public isEditable?: boolean;
  public key!: string;
  public search?: SearchFunc;
  public sortDirection?: SortDirection;
  public style?: React.CSSProperties;
  public title?: string;
  public validation?: ValidationFunc;
}
