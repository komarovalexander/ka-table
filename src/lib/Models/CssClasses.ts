export class CssClasses {
  public table?: string = 'ka-table';
  public tbody?: string = 'ka-tbody';
  public thead?: string = 'ka-thead';
  public theadRow?: string = 'ka-tr ka-thead-row';
  public theadCell?: string = 'ka-thead-cell';
  public groupRow?: string = 'ka-tr ka-group-row';
  public row?: string = 'ka-tr ka-row';
  public rowSelected?: string = 'ka-row-selected';
  public cell?: string = 'ka-cell';

  public iconGroupArrowCollapsed?: string = 'ka-icon ka-icon-group-arrow ka-icon-group-arrow-collapsed';
  public iconGroupArrowExpanded?: string = 'ka-icon ka-icon-group-arrow ka-icon-group-arrow-expanded';
  public iconClose?: string = 'ka-icon ka-icon-close';
  public iconSortArrowDown?: string = 'ka-icon ka-icon-sort ka-icon-sort-arrow-down';
  public iconSortArrowUp?: string = 'ka-icon ka-icon-sort ka-icon-sort-arrow-up';

  public checkbox ?: string = 'ka-input';
  public dateInput ?: string = 'ka-input';
  public numberInput ?: string = 'ka-input';
  public textInput ?: string = 'ka-input';
}
