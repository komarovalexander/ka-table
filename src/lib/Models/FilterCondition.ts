/* Filter condition for specific field */
export class FilterCondition {
  public field!: string;
  public operator!: string;
  public value!: any;
}
