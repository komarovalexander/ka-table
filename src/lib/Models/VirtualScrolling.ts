export class VirtualScrolling {
  public scrollPosition?: number;
  public itemHeight?: (data: any) => number | number;
  public tbodyHeight?: number;
}
