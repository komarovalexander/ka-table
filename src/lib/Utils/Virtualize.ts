import { VirtualScrolling } from '../Models/VirtualScrolling';

export const getVirtualized = (virtualScrolling: VirtualScrolling, data: any[]) => {
  const virtualizedData: any[] = [];
  const { scrollPosition = 0 } = virtualScrolling;
  let { tbodyHeight = 600 } = virtualScrolling;
  let beginHeight = 0;
  let endHeight = 0;
  data.reduce((acc, value) => {
    const itemHeight = virtualScrolling.itemHeight ?
      (
        typeof virtualScrolling.itemHeight === 'number'
        ? virtualScrolling.itemHeight
        : virtualScrolling.itemHeight(value)
      )
      : 40;
    if (acc >= scrollPosition - itemHeight) {
      if (tbodyHeight >= -(itemHeight * 5)) {
        tbodyHeight = tbodyHeight - itemHeight;
        virtualizedData.push(value);
      } else {
        endHeight += itemHeight;
      }
    } else {
      beginHeight = acc - itemHeight;
    }
    return acc + itemHeight;
  }, 0);

  return {
    beginHeight,
    endHeight,
    virtualizedData,
  };
};
