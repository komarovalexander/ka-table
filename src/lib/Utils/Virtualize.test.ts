import { VirtualScrolling } from '../Models/VirtualScrolling';
import { getVirtualized } from './Virtualize';

describe('Virtualize', () => {
  const data: any[] = new Array(100).fill(0).map((_, index) => index);
  describe('getVirtualized', () => {
    it('scrollPosition 0', () => {
      const virtualScrolling: VirtualScrolling = {
        itemHeight: () => 10,
        scrollPosition: 0,
        tbodyHeight: 40,
      };
      const result = getVirtualized(virtualScrolling, data);

      expect(result).toMatchSnapshot();
    });

    it('scrollPosition 100', () => {
      const virtualScrolling: VirtualScrolling = {
        itemHeight: () => 10,
        scrollPosition: 100,
        tbodyHeight: 40,
      };
      const result = getVirtualized(virtualScrolling, data);

      expect(result).toMatchSnapshot();
    });

    it('scrollPosition 900', () => {
      const virtualScrolling: VirtualScrolling = {
        itemHeight: () => 10,
        scrollPosition: 900,
        tbodyHeight: 40,
      };
      const result = getVirtualized(virtualScrolling, data);

      expect(result).toMatchSnapshot();
    });

    it('itemHeight 40', () => {
      const virtualScrolling: VirtualScrolling = {
        itemHeight: () => 40,
        scrollPosition: 0,
        tbodyHeight: 160,
      };
      const result = getVirtualized(virtualScrolling, data);

      expect(result).toMatchSnapshot();
    });
  });
});
