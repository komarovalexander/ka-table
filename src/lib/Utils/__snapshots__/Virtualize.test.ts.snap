// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Virtualize getVirtualized itemHeight 40 1`] = `
Object {
  "beginHeight": 0,
  "endHeight": 3600,
  "virtualizedData": Array [
    0,
    1,
    2,
    3,
    4,
    5,
    6,
    7,
    8,
    9,
  ],
}
`;

exports[`Virtualize getVirtualized scrollPosition 0 1`] = `
Object {
  "beginHeight": 0,
  "endHeight": 900,
  "virtualizedData": Array [
    0,
    1,
    2,
    3,
    4,
    5,
    6,
    7,
    8,
    9,
  ],
}
`;

exports[`Virtualize getVirtualized scrollPosition 100 1`] = `
Object {
  "beginHeight": 70,
  "endHeight": 810,
  "virtualizedData": Array [
    9,
    10,
    11,
    12,
    13,
    14,
    15,
    16,
    17,
    18,
  ],
}
`;

exports[`Virtualize getVirtualized scrollPosition 900 1`] = `
Object {
  "beginHeight": 870,
  "endHeight": 10,
  "virtualizedData": Array [
    89,
    90,
    91,
    92,
    93,
    94,
    95,
    96,
    97,
    98,
  ],
}
`;
